lib/HLSL/DxcOptimizer.cpp

```diff
@@ -221,6 +225,69 @@ HRESULT STDMETHODCALLTYPE DxcOptimizer::GetAvailablePass(
       GetPassArgDescriptions(m_passes[index]->getPassArgument()), ppResult);
 }
 
+template <typename _T>
+static void CopyResourceInfo(_T &TargetRes, const _T &SourceRes,
+                             DxilTypeSystem &TargetTypeSys,
+                             const DxilTypeSystem &SourceTypeSys) {
+  if (TargetRes.GetKind() != SourceRes.GetKind() ||
+      TargetRes.GetLowerBound() != SourceRes.GetLowerBound() ||
+      TargetRes.GetRangeSize() != SourceRes.GetRangeSize() ||
+      TargetRes.GetSpaceID() != SourceRes.GetSpaceID()) {
+    DXASSERT(false, "otherwise, resource details don't match");
+    return;
+  }
+
+  if (TargetRes.GetGlobalName().empty() && !SourceRes.GetGlobalName().empty()) {
+    TargetRes.SetGlobalName(SourceRes.GetGlobalName());
+  }
+
+  Type *Ty = SourceRes.GetHLSLType();
+  TargetRes.SetHLSLType(Ty);
+
+  if (isa<PointerType>(Ty))
+    Ty = Ty->getPointerElementType();
+
+  while (isa<ArrayType>(Ty))
+    Ty = Ty->getArrayElementType();
+
+  // TODO: Detect if struct annotation is stripped, then remove it if so, so
+  // it can be replaced.  For now, just always replace it.
```

Comment:
I don't know if this comment is very clear, and I don't think the case is necessary after all.  I'll remove the comment.